daemon/containerd/image_list.go

```diff
@@ -138,7 +111,73 @@ func (i *ImageService) Images(ctx context.Context, opts types.ImageListOptions)
 	return summaries, nil
 }
 
-type imageFilterFunc func(image containerd.Image) bool
+func (i *ImageService) singlePlatformImage(ctx context.Context, img images.Image, platform v1.Platform, opts types.ImageListOptions) (*types.ImageSummary, []digest.Digest, error) {
+	platMC := c8dplatforms.OnlyStrict(platform)
+	if avail, _, _, _, err := images.Check(ctx, i.client.ContentStore(), img.Target, platMC); !avail {
```

Comment:
avail can be true in case the manifest is present in the content store, but some of its children is missing.
In this case, and if the Check returns an error, we probably want to do an early return:
```suggestion
	avail, _, _, missing, err := images.Check(ctx, i.client.ContentStore(), img.Target, platMC)
	if !avail || err != nil || len(missing) > 0 {
```

Reply:
Yes. The doc is a bit misleading, one has to read it multiple times to understand it correctly 😄 

If available is true, the caller can assume that `required` (one of return values) represents the complete set of content required for the image.

And the latter fragment:
`missing` will have the components that are part of `required` but not available in the provider.

The code also confirms:
https://github.com/containerd/containerd/blob/main/images/image.go#L323

Reply:
Yeah ok looks like the documentation is not right

Reply:
`missing` will contain the descriptor of the rootfs but I can run the image just fine (with `docker run --platform linux/amd64 hello-world`)...

Reply:
That's strange, I'm getting different results. Pulling amd64 starting from fresh state, yields these results for me:
```
# docker pull --platform linux/amd64 hello-world
```

```
available=true error="<nil>" missing="[]" platform="{amd64 linux  [] }"
available=false error="<nil>" missing="[{application/vnd.docker.distribution.manifest.list.v2+json sha256:faa03e786c97f07ef34423fccceeec2398ec8a5759259f94d99078f264e9d7af 2561 [] map[] [] <nil>}]" platform="{arm linux  [] v5}"
available=false error="<nil>" missing="[{application/vnd.docker.distribution.manifest.list.v2+json sha256:faa03e786c97f07ef34423fccceeec2398ec8a5759259f94d99078f264e9d7af 2561 [] map[] [] <nil>}]" platform="{arm linux  [] v7}"
available=false error="<nil>" missing="[{application/vnd.docker.distribution.manifest.list.v2+json sha256:faa03e786c97f07ef34423fccceeec2398ec8a5759259f94d99078f264e9d7af 2561 [] map[] [] <nil>}]" platform="{arm64 linux  [] v8}"
...

```

Then doing:
```
# echo 'FROM hello-world' | docker buildx build -t asdf -
```

will produce:
```
available=true error="<nil>" missing="[]" platform="{amd64 linux  [] }"
available=true error="<nil>" missing="[{application/vnd.docker.image.rootfs.diff.tar.gzip sha256:b921b04d0447ddcd82a9220d887cd146f6ef39e20a938ee5e19a90fc3323e030 3684 [] map[] [] <nil>}]" platform="{arm linux  [] v5}"
available=true error="<nil>" missing="[{application/vnd.docker.image.rootfs.diff.tar.gzip sha256:9b157615502ddff86482f7fe2fa7a074db74a62fce12b4e8507827ac8f08d0ce 2993 [] map[] [] <nil>}]" platform="{arm linux  [] v7}"
available=true error="<nil>" missing="[]" platform="{arm64 linux  [] v8}"
...
```

so images ls will show multiple images:
```
# docker image ls
REPOSITORY    TAG       IMAGE ID       CREATED         SIZE
asdf          latest    75f42e558e3d   7 seconds ago   4.63kB
hello-world   latest    faa03e786c97   3 minutes ago   7.03kB
hello-world   latest    faa03e786c97   3 minutes ago   8.25kB
hello-world   latest    faa03e786c97   3 minutes ago   7.56kB
hello-world   latest    faa03e786c97   3 minutes ago   7.78kB
```

.. which are not available locally:
```
# docker run --rm --platform linux/arm/v5 hello-world
Unable to find image 'hello-world:latest' locally
^C
```